### DISCUSSION TITLE
BLE Support

### DISCUSSION THREAD

**Discussion by vovagorodok** [JAndrassy/ArduinoOTA#157] [General]
Any plans to add support OTA over BLE?
Found this one: https://github.com/fbiego/ESP32_BLE_OTA_Arduino
Will be nice to have here such functionality

**Comment by JAndrassy** [comment]
> Any plans to add support OTA over BLE? Found this one: https://github.com/fbiego/ESP32_BLE_OTA_Arduino Will be nice to have here such functionality

just use InternalStorage object from this library instead of the esp32 Update object

**Reply by vovagorodok** [reply]
Decided to create Arduino library that base on ArduinoOTA:
https://github.com/vovagorodok/ArduinoBleOTA

In order to use `OTAStorage` is needed to copy all includes:
```
#ifdef __AVR__
#if FLASHEND >= 0xFFFF
#include "InternalStorageAVR.h"
#endif
#elif defined(STM32F1xx)
#include <InternalStorageSTM32.h>
#elif defined(ARDUINO_ARCH_RP2040)
#include <InternalStorageRP2.h>
#elif defined(ESP8266) || defined(ESP32)
#include "InternalStorageESP.h"
#else
#include "InternalStorage.h"
#endif
#ifdef __SD_H__
#include "SDStorage.h"
SDStorageClass SDStorage;
#endif
#ifdef SerialFlash_h_
#include "SerialFlashStorage.h"
SerialFlashStorageClass SerialFlashStorage;
#endif
```
here: https://github.com/vovagorodok/ArduinoBleOTA/blob/main/src/OTAStorage.h

Can that includes be moved to separate include file? (it helps to easy maintain BleOTA)

**Comment by JAndrassy** [comment]
you would have to copy everything except of ArduinoOTA.h and WiFiOTA.cpp/.h.
better use the ArduinoOTA library as dependency. 
put 
```
depends=ArduinoOTA
```
into library.properties of your library

**Reply by vovagorodok** [reply]
Currently it's arduino program instead of library. I use ArduinoOTA as dependency:
https://github.com/vovagorodok/ArduinoBleOTA/blob/main/platformio.ini
Nothing is copied except lines with defines of storage.

But in order to have the same interface as ArduinoOTA, I need this includes. When I include ArduinoOTA.h, Wifi library is included to. Will be nice create separate  `OTAStorageDefinitions.h` with content:
```
#pragma once

#ifdef __AVR__
#if FLASHEND >= 0xFFFF
#include "InternalStorageAVR.h"
#endif
#elif defined(STM32F1xx)
#include <InternalStorageSTM32.h>
#elif defined(ARDUINO_ARCH_RP2040)
#include <InternalStorageRP2.h>
#elif defined(ESP8266) || defined(ESP32)
#include "InternalStorageESP.h"
#else
#include "InternalStorage.h"
#endif
#ifdef __SD_H__
#include "SDStorage.h"
SDStorageClass SDStorage;
#endif
#ifdef SerialFlash_h_
#include "SerialFlashStorage.h"
SerialFlashStorageClass SerialFlashStorage;
#endif
```

and include it in `ArduinoOTA.h`:
```
#ifndef _ARDUINOOTA_H_
#define _ARDUINOOTA_H_

#include "WiFiOTA.h"
#include "OTAStorageDefinitions.h"

const uint16_t OTA_PORT = 65280;

template <class NetServer, class NetClient>
class ArduinoOTAClass : public WiFiOTAClass {
...
```

and I'll include only storage definitions without wifi in `BleOTA.h`:
```
#pragma once
#include <NimBLEDevice.h>
#include "OTAStorageDefinitions.h"

class BleOTAClass: public BLECharacteristicCallbacks
{
...
```

**Comment by JAndrassy** [comment]
have OTAStorageDefinitions.h in your library.

**Reply by vovagorodok** [reply]
It 'will be copy-pasta of ArduinoOTA code. I would avoid it, if it's possible

**Reply by JAndrassy** [reply]
it is not a problem since maybe it will not even be the same set of supported platforms

**Reply by vovagorodok** [reply]
But if something will changed in OTAStorageDefinitions of ArduinoOTA repo, I should align BleOta repo each time (additional maintenance). Better to support all platforms from internal storage PoV, and have it abstract, it helps easily extend BleOta in future

**Comment by EMP222** [comment]
Does this library support Arduino Nano 33 BLE rev2 (nRF52840)?
